### ISSUE TITLE
Kafka Streams 1.3.0 Dependencies Not Resolving

### ISSUE THREAD

**Issue by ctoestreich** [micronaut-projects/micronaut-kafka#83]
Trying the new 1.3.0.M1 which includes the new kafka streams 1.3.0.  Used a simple generate project and not getting unresolved dependencies.

* Cleared m2 no dice
* Reverting microanut-kafka-streams to 1.2.0 fixes this issue
* LOOKS like the issue here is that the dep scope changed to runtime

Example project which is broken: https://github.com/tardisoft/kstream-1.3.0

![Screen Shot 2019-11-25 at 10 48 13 AM](https://user-images.githubusercontent.com/691357/69560481-71038a80-0f71-11ea-82a0-dc32a81e4f46.png)

<img width="1813" alt="Screen Shot 2019-11-25 at 10 53 32 AM" src="https://user-images.githubusercontent.com/691357/69560781-e2433d80-0f71-11ea-8dcb-40232adf259e.png">

**Comment by Sabst** [comment]
I got stuck on that too. Thanks @ctoestreich  for reporting this.

Another reproducer is just to create an empty app using kafka-streams:
`mn create-app test131rc1 -f kafka-streams`
And include any kafka class in `Application.groovy`, say:
`import org.apache.kafka.clients.producer.internals.DefaultPartitioner`

I wonder how any test can pass...

BTW, also had to get rid of `afterEvaluate` in the `publish.gradle` file to allow the publication of 1.3.1 to local maven (that used to worked well with 1.2.0 and before)... but micronaut-kafka 1.3.1 has the dependencies problem too.

**Comment by graemerocher** [comment]
This I believe is not a bug in this project but a change in the way Gradle exposes dependencies. Prior to Gradle 5 `compile` dependencies where included in `compile` scope and now they are `runtime` in the POM. Not sure what we can do about it

**Comment by ctoestreich** [comment]
@graemerocher The root cause might not be a "bug" in Micronaut, but clearly this is still a bug as the generated poms for Micronaut are breaking.  I am not familiar with Gradle 5 changes as I have been stuck on maven for a long time.  Is there any accommodation you can make for these scope changes when the pom files are created?  I haven't dug too deep but there is clearly some gradle->pom.xml magic in this file https://raw.githubusercontent.com/micronaut-projects/micronaut-build/3ca806eb776b72483a4c0fda5ff1a5caebe0c281/publishing.gradle

**Comment by Sabst** [comment]
@graemerocher I checked with Micronaut 1.3.0.M2 and the problem described above is still there.
Micronaut Kafka is unusable and I see nothing in the release notes or documentation that explains this change.

This has nothing to do with Gradle 5 as we made this move a while back. Also, what I proposed as a reproducer above (`mn create-app ...`) uses the gradle version recommended by Micronaut.

I managed to progress by adding these additional dependencies to 1.3 applications (I took them from the Micronaut Kafka source code):
```
    implementation "org.apache.kafka:kafka-clients:2.3.1"
    implementation "org.apache.kafka:kafka-streams:2.3.1"

    compileOnly "org.apache.kafka:kafka_2.12:2.3.1"
    compileOnly "org.apache.kafka:kafka-clients:2.3.1:test"
    compileOnly "org.apache.kafka:kafka_2.12:2.3.1:test"
```
As you see, something is broken in Micronaut Kafka 1.3 itself...

As a side note, even with the "workaround" above, the application still fails due to an issue that looks like #87... Can it be related to the dubious build and runtime conditions? What I see is that some listeners are silently ignored.

**Comment by ctoestreich** [comment]
I haven’t tried 1.3.0.M2 but with M1 I am using the same work around in maven.

**Comment by Sabst** [comment]
Ok, maybe some slight progress... the compilation/dependency issue is still unclear (I'm still using the workaround above) but the total failure of the running app (is it the same problem as #87?) seems to be due to this missing dependency:
`implementation "io.micronaut:micronaut-messaging:1.3.0.M2"` (adapt the version as needed)

Please confirm if you have this problem (e.g. no more consumer available) with Kafka and Micronaut 1.3.

If this new dependency is now required, this will require at least a documentation and kafka feature (the dependency is not added when creating an app with `-f kafka-streams`) update.

It is also possible that this dependency was implicit in earlier Micronaut versions and that this is yet another effect of the broken "transitive dependency resolution" that has been already seen with the kafka dependencies above...

**Comment by graemerocher** [comment]
We have been recently moving Micronaut to Gradle 6 and the new java library stuff. This seems to have broken that dependency somehow. Will investigate the cause

**Comment by graemerocher** [comment]
Can you try the `1.4.0.BUILD-SNAPSHOT` version when the build passes, the use of `api` instead of `implementation` changes the scope of the dependencies on the POM from `runtime` to `compile`

**Comment by ctoestreich** [comment]
@graemerocher what is the snapshot repo you guys push to?

I have this and none of these worked.

```xml
<repositories>
        <repository>
            <id>bintray2</id>
            <name>bintray2</name>
            <url>http://oss.jfrog.org/artifactory/oss-snapshot-local/</url>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
        <repository>
            <id>bintray</id>
            <name>bintray</name>
            <url>https://jcenter.bintray.com</url>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
    </repositories>
```

**Comment by graemerocher** [comment]
The snapshots can be found here https://oss.jfrog.org/artifactory/oss-snapshot-local/io/micronaut/configuration/micronaut-kafka-streams/

**Comment by Sabst** [comment]
@graemerocher Some update from my side...
My apps are working again using Micronaut 1.3 and the snapshots :tada:
I also managed to move the tests to Test Containers.

What I see is that the impact of each Micronaut version is quite strong and not every project can afford spending this migration/investigation time. It's a pain point that should be addressed somehow.

Also, this issue shows that the per-module testing is not enough: obviously, the kafka module tests showed things were ok when they were not from an external application standpoint.
In addition, examples of migration would help a lot when it is decided to move from technology A to B (say Kafka embedded to Test Containers).

IMO, what is missing is
- a reference application (or applications) that uses as many features/modules as possible and that helps the validation (by Micronaut Core Team or any contributor) of new developments from an external standpoint (e.g. a petstore++)
- a migration documentation (release notes are nice but it's - most of the time - hard to see what the implications of the changes are) that could be created while porting this application to the new Micronaut version

**Comment by ctoestreich** [comment]
@Sabst we should help adding some containerized tests or some external component tests for the Kafka code. We do this in our enterprise stack currently at work. 

The issue here has do do with also having a variance of build systems and micro auto doesn’t have a full suite of maven tests for every config repo. 

Maybe we can collaborate a bit to help find a good solution to bridge both issues.

**Comment by graemerocher** [comment]
@Sabst to be fair 1.3 is at a milestone stage which means it is unlikely to be perfect and the the reason we release milestones is to catch issues like this such that when 1.3 GA comes out the upgrade is as easy as possible. We appreciate greatly that folks you like and @ctoestreich are trying the milestones on 1.3 so that we can fix issues that come up prior to release.

Generally with Micronaut 1.x we are trying to following semantic versioning so there should be as little breaking changes and upgrade problems as possible. That doesn't mean problems don't occur and maybe we could do more to test individual modules like the Kafka one against the snapshots of 1.3 so thanks for the feedback there we will see what we can do.

**Comment by Sabst** [comment]
@ctoestreich Yes this could be a contribution/PR too. I'm a one-person company and I cannot contribute as much as I would like. My focus is more about making sure I can safely push for this technology when I finally meet a Customer ready for the ride, especially here in France where I do not see that many Micronauts (but I'm used to it as a Grails fan!). One of the tricky parts is usually to move from one version to another and I know companies that were stuck on Grails 2 for a long time.

@graemerocher Yes, this was meant as a constructive comment and I am really grateful for the path you/the Core Team are following (including the organic link to Grails) and all you have achieved so far! It's just that I like to see each issue as an opportunity for root cause analysis... sorry about that!
My point was not so much about more individual module testing but more about a "framework validation showcase" (one or several applications) that would put those who work on the framework internals more in the end-users shoes. And then share this experience with the actual end-users in some migration documentation.
In a nutshell: still allow innovation and breaking changes but mitigate the associated friction
